File: Morhun_transfusion_task.pl
% Задаємо ємність посудин
capacity([8, 3, 7]). 

% Задаємо початковий стан і фінальний стан
start_stage([8, 0, 0]).
goal_stage([3, 0, 5]).

% Операція переливання води з однієї посудини в іншу
pour([X, Y, Z], [X1, Y1, Z]) :- capacity([_, MaxY, _]), transfer(X, Y, X1, Y1, MaxY).
pour([X, Y, Z], [X1, Y, Z1]) :- capacity([_, _, MaxZ]), transfer(X, Z, X1, Z1, MaxZ).
pour([X, Y, Z], [X, Y1, Z1]) :- capacity([_, _, MaxZ]), transfer(Y, Z, Y1, Z1, MaxZ).
pour([X, Y, Z], [X1, Y1, Z]) :- capacity([MaxX, _, _]), transfer(Y, X, Y1, X1, MaxX).
pour([X, Y, Z], [X1, Y, Z1]) :- capacity([MaxX, _, _]), transfer(Z, X, Z1, X1, MaxX).
pour([X, Y, Z], [X, Y1, Z1]) :- capacity([_, MaxY, _]), transfer(Z, Y, Z1, Y1, MaxY).

% Функція переливаємо з однієї посудини в іншу
transfer(A, B, A1, B1, MaxB) :-  
    Total is A + B,  
    (Total =< MaxB -> A1 = 0, B1 = Total ; A1 is Total - MaxB, B1 = MaxB).

% Основний пошук шляху
path(P) :-
    start_stage(Start),
    goal_stage(Goal),
    bfs_algorithm([[Start]], Goal, RevPath),
    reverse(RevPath, P).              % отримуємо список у зворотньому порядку, 
                                      % бо кожне переливання додається на початок списку

% Алгоритм пошуку в ширину
bfs_algorithm([[Goal | Rest] | _], Goal, [Goal | Rest]).
bfs_algorithm([[State | Rest] | Queue], Goal, P) :-
    findall([Next, State | Rest],
            (pour(State, Next), \+ member(Next, [State | Rest])),
            NewPaths),
    append(Queue, NewPaths, NewQueue),
    bfs_algorithm(NewQueue, Goal, P).

File: Morhun_transfusion_task.hs
import Data.List (nub)
import qualified Data.Set as Set

type State = (Int, Int, Int)

capacity :: (Int, Int, Int)
capacity = (9, 4, 7)

startStage :: State
startStage = (9, 4, 0)

goalStage :: State
goalStage = (6, 4, 3)

-- Функція переливання
pour :: State -> [State]
pour (x, y, z) = nub $ filter (`Set.notMember` visited) newStates
  where
    (maxX, maxY, maxZ) = capacity
    newStates = [
        transfer (x, y) maxY (\x' y' -> (x', y', z)),
        transfer (x, z) maxZ (\x' z' -> (x', y, z')),
        transfer (y, x) maxX (\y' x' -> (x', y', z)),
        transfer (y, z) maxZ (\y' z' -> (x, y', z')),
        transfer (z, x) maxX (\z' x' -> (x', y, z')),
        transfer (z, y) maxY (\z' y' -> (x, y', z'))
      ]
    visited = Set.empty  -- уникаємо повторень

-- Переливаємо воду між двома ємностями
transfer :: (Int, Int) -> Int -> (Int -> Int -> State) -> State
transfer (a, b) maxB f
  | a + b <= maxB = f 0 (a + b)
  | otherwise     = f (a - (maxB - b)) maxB

-- Пошук у ширину (BFS)
bfs :: [[State]] -> Set.Set State -> [State]
bfs [] _ = []
bfs (path@(current:_):rest) visited
  | current == goalStage = reverse path
  | otherwise =
      let newStates = filter (`Set.notMember` visited) (pour current)
          newPaths = map (:path) newStates
          newVisited = foldr Set.insert visited newStates
      in bfs (rest ++ newPaths) newVisited

solve :: [State]
solve = bfs [[startStage]] Set.empty

main :: IO ()
main = print solve
